fix: resolve peer dependencies not updated with `pnpm upgrade --latest`

m2na7 opened the pull request:
fixes #9900

Fix peer dependencies not being upgraded when using `pnpm upgrade --latest`.

## Problem

In monorepos, peer dependencies (both catalog-based and regular) were not being upgraded when running `pnpm upgrade --latest`, while regular dependencies updated correctly.

**Root cause**: The `--latest` flag triggers the `installSome` code path, which uses `getAllDependenciesFromManifest()` to build `currentBareSpecifiers`. Unlike `installCase` which uses `getWantedDependencies()` and includes peer dependencies, `getAllDependenciesFromManifest()` was excluding them entirely.

## Solution

Added an optional `autoInstallPeers` parameter to `getAllDependenciesFromManifest` to make it behave consistently with `getWantedDependencies`:

  ```typescript
  export function getAllDependenciesFromManifest (
    manifest: Pick<ProjectManifest, DependenciesField | 'peerDependencies'>,
    opts?: { autoInstallPeers?: boolean }
  ): Dependencies {
    return {
      ...manifest.devDependencies,
      ...manifest.dependencies,
      ...manifest.optionalDependencies,
      ...(opts?.autoInstallPeers ? manifest.peerDependencies : {}),
    }
  }
```

Now installSome and installCase handle peer dependencies consistently.

## Key Changes

1) Modified getAllDependenciesFromManifest to accept autoInstallPeers option
2)  Updated installSome to pass autoInstallPeers when calling getAllDependenciesFromManifest
3) Added test for peer dependency upgrades in install/update.ts

m2na7 commented:
@zkochan I’ve updated the logic and added tests to ensure the issue is properly covered. 
Please have a look at the PR whenever you’re available!

LaurentChardin commented:
Any news on this one ? it doesn't seem to follow `auto-install-peers=true` in `.npmrc` neither.

gluxon commented on pkg-manager/resolve-dependencies/src/index.ts:
```diff
@@ -281,10 +283,10 @@ export async function resolveDependencies (
     const resolvedImporter = resolvedImporters[project.id]
     for (let i = 0; i < resolvedImporter.directDependencies.length; i++) {
-      if (project.wantedDependencies[i]?.updateSpec == null) continue
+      if (project.wantedDependencies[i]?.updateSpec !== true) continue
```
This is definitely a bug. Thanks for fixing this part.

gluxon replied:
Would appreciate a separate PR for this one if you get a chance! I think it's unrelated to the peer dependencies bug, but still worth a fix in a separate PR.

gluxon commented on pkg-manager/core/src/install/index.ts:
```diff
@@ -563,7 +563,12 @@ export async function mutateModules (
+        : {
+          ...getAllDependenciesFromManifest(project.manifest),
+          ...project.manifest.peerDependencies,
+        }
```
This fix doesn't feel right since we're changing code that affects every dependency regardless of whether it's using the catalog. Could you either:

1. Give more details on why this would be the right fix, or
2. Look into why this only happens for peer dependencies using the `catalog:` protocol, but works fine when it doesn't use the `catalog:` protocol?

Since peer dependency upgrades work without using the `catalog:` protocol, I wouldn't expect code in the general case to need to be modified here.

I also think there may be a bug here — if we do go with this fix (which I'm not sure is right), we would need to respect the `auto-install-peers` flag. Peer dependencies in workspace packages aren't installed if that flag is false.

gluxon replied:
In my first comment, I thought it was strange that we were applying a general fix for a catalog-specific bug. What I'm now realizing after creating a secondary repro today is that this bug is completely independent of the `catalog:` protocol.

https://github.com/gluxon/pnpm-issue-9900-repro

In the repro above, I declared `peerDependencies` entries on one dependency using the catalog, and another without. Neither entry is updated on `pnpm update --latest -r`. I can even remove the catalog entries entirely from `pnpm-workspace.yaml` entirely and the issues still occur.

Given that new understanding, I think we should go back to targeting a general fix since this issue isn't related to catalogs.

gluxon replied:
What the original bug report (https://github.com/pnpm/pnpm/issues/9900) did get right though is that this only happens with the `--latest` flag. I think there's a bug where that flag is changing what dependencies are in scope for the upgrade that we need to fix.

gluxon replied:
I'm now following why you had to modify the `installSome` case, but I think we were missing an explanation of why that had to be modified. That part wasn't clear until I spun up a debugger and saw that the is that the `--latest` flag causes that code path to be hit. Without the `--latest` flag, we fall into the `installCase` branch instead, which does handle peer dependencies correctly.

https://github.com/pnpm/pnpm/issues/9900#issuecomment-3708384805

m2na7 replied:
@gluxon Completely agree! Thank you for creating the reproduction.

  As you suggested, I've removed the catalog filter and now include all peer
  dependencies. It now works the same way as installCase.

```ts
    async function installSome (project: InstallSomeProject) {
      const currentBareSpecifiers = opts.ignoreCurrentSpecifiers
        ? {}
        : {
          ...getAllDependenciesFromManifest(project.manifest),
          ...(opts.autoInstallPeers ? project.manifest.peerDependencies ?? {} : {}),
        }

```
Now installSome and installCase handle peer dependencies consistently.

gluxon reviewed (commented):
Thanks for the help @m2na7! I realize it's been some time since the PR was opened and when this got reviewed, so if you're unsure of the answer to any questions I had, that's okay.

gluxon commented on pkg-manager/resolve-dependencies/src/index.ts:
I suspect we can revert the changes here? I think the real fix is making the `installSome` code path line up with the behavior in the `installCase` code path and the changes here are superfluous.

gluxon commented on pkg-manager/core/test/catalogs.ts:
```diff
@@ -1540,6 +1540,89 @@ describe('update', () => {
     expect(Object.keys(lockfile.snapshots)).toEqual(['@pnpm.e2e/foo@1.3.0'])
   })
 
+  test('pnpm upgrade --latest -r should update both dependencies and peerDependencies using catalog', async () => {
```
Since this issue isn't related to catalogs, could this test be moved to `pkg-manager/core/test/install/update.ts`?

gluxon commented on pkg-manager/resolve-dependencies/src/index.ts:
```diff
@@ -281,10 +283,10 @@ export async function resolveDependencies (
   let updatedCatalogs: Record<string, Record<string, string>> | undefined
   for (const project of projectsToResolve) {
-    if (!project.updatePackageManifest) continue
+    if (!project.updatePackageManifest && !project.updateToLatest) continue
```
> Only the catalog needs to be updated, while package.json remains unchanged (keeps `"catalog:"`)

This is the part that I think isn't fully correct. Suppose updates happen to dependencies using the catalog protocol and dependencies not using the catalog protocol. It wouldn't be correct to only update the catalog protocol specifiers in `pnpm-workspace.yaml` and not update `package.json`.

The `updatePackageManifest` field isn't named as well as it could be at the moment. It currently means whether both `pnpm-workspace.yaml` and `package.json` should be updated.

m2na7 replied:
You're right that updatePackageManifest means updating both.
I've reverted these changes. The core fix in installSome is sufficient.

gluxon reviewed (approved):
Thanks @m2na7! This looks good to me. Let me pass this to @zkochan for the final review.

gluxon commented:
@zkochan I think this fix is ready for review and should fix some surprises when using `pnpm update` that people like @m2na7 and @michkot are seeing. Thanks.

zkochan commented:
sorry, I missed this

zkochan committed:
fix: upgrade peer dependencies with `pnpm upgrade --latest`

The `--latest` flag triggers the `installSome` code path, which built
`currentBareSpecifiers` via `getAllDependenciesFromManifest()` — a
function that excluded peer dependencies. The non-`--latest` path uses
`getWantedDependencies()`, which honors `autoInstallPeers` and includes
them. Pass `autoInstallPeers` through `getAllDependenciesFromManifest`
so both paths agree.

Closes #9900
